docs: Update README.adoc (#2800)
dfinity/ic: rs/p2p/quic_transport/README.adoc

== Requirements ==

[cols="3,3,3"]
|===
| **Capability**
| **Purpose**
| **Implementation Requirement**

| RPC-like library
| Simplifies application development by abstracting communication between nodes, allowing developers to focus on application logic instead of low-level networking.
| Ensure https://en.wikipedia.org/wiki/Reliability_(computer_networking)[reliable data delivery] so the sender can confirm successful message delivery to the recipient.

> | Multitenancy of applications

Reviewer: ```suggestion
| Multitenancy of applications.
```

Author: I would prefer to keep it without the dot because it is more of a title than a full sentence

| Enables multiple applications to safely share the library, ensuring isolation and uninterrupted operation across applications.
| Support https://en.wikipedia.org/wiki/Multiplexing[multiplexing] to independently route messages to their endpoints, unaffected by packet losses or delays impacting other messages.

| Operates in P2P networks with Byzantine peers
| Supports deployment in real-world environments, including networks with faulty or malicious nodes.
| Implement https://en.wikipedia.org/wiki/Flow_control_(data)[flow control] to manage the number of connections, inflight messages, and streams per connection, ensuring stability and resilience.

> | Scalable and resource-efficient

Reviewer: ```suggestion
| Scalable and resource-efficient.
```

Author: ditto

Job execution starts in the threadpool iff the parent task was not cancelled.
A nice writeup about async and blocking operations can be found in https://ryhl.io/blog/async-what-is-blocking/[Alice Ryhl's blog post].

== Implementation Details ==

> Use the https://github.com/quinn-rs/quinn[QUINN] library for QUIC, as it meets the requirements for reliable data delivery, multiplexing, connection-oriented communication, and flow control.  

Reviewer: ```suggestion
This RPC library uses the https://github.com/quinn-rs/quinn[QUINN] library for QUIC, as it meets the requirements for reliable data delivery, multiplexing, connection-oriented communication, and flow control.  
```

=== QUIC stream per message ===

Instead of one stream per handler/URI. TODO: explain.

=== Handlers are always ready to process messages ===

> The router is always ready to accept streams. TODO: explain.

Reviewer: Would be good to create a Jira ticket for these.

Author: https://dfinity.atlassian.net/browse/NET-1779
